### PR TITLE
zmq: Add support to listen on multiple interfaces

### PR DESCRIPTION
**Pull request by n-thumann**
This PR adds support for ZeroMQ to listen on multiple interfaces, just like the RPC server.
Currently, if you specify more than one e.g. `zmqpubhashblock` paramter, only the first one will be used. Therefore a user may be forced to listen on all interfaces (e.g. `zmqpubhashblock=0.0.0.0:28332`), which can result in an increased attack surface.
With this PR a user can specify multiple interfaces to listen on, e.g.
`-zmqpubhashblock=tcp://127.0.0.1:28332 -zmqpubhashblock=tcp://192.168.1.123:28332`.

### CODE REVIEW

**Review by promag — COMMENTED** [review]
Concept ACK.

Maybe use a minimal diff
```diff
     for (const auto& entry : factories)
     {
         std::string arg("-zmq" + entry.first);
-        if (gArgs.IsArgSet(arg))
-        {
-            CZMQNotifierFactory factory = entry.second;
-            std::string address = gArgs.GetArg(arg, "");
+        CZMQNotifierFactory factory = entry.second;
+        for (const std::string& address : gArgs.GetArgs(arg)) {
             CZMQAbstractNotifier *notifier = factory();
             notifier->SetType(entry.first);
             notifier->SetAddress(address);
```
Also:
 - update the doc/zmq.md with this change
 - add release note too
 - update some test in test/functional/interface_zmq.py with multiple interfaces.

Welcome!

**Comment by promag** [comment]
> * update some test in test/functional/interface_zmq.py with multiple interfaces.

@n-thumann don't forget about this.

**Comment by n-thumann** [comment]
> > * update some test in test/functional/interface_zmq.py with multiple interfaces.
> 
> @n-thumann don't forget about this.

Thank you for your feedback! The test is already in the works :)

**Comment by n-thumann** [comment]
Currently the `debug.log` prints
`2020-03-11T17:52:01.715152Z [scheduler] zmq: Publish hashblock <blockhash>
2020-03-11T17:52:01.715248Z [scheduler] zmq: Publish hashblock <blockhash>`, when `hashblock` is triggered and ZeroMQ is listening on two interfaces.
Do you think it´s useful to append something like `to tcp://127.0.0.1:28332` to each log entry, so that a user knows why there is an apparently duplicate entry?

**Comment by promag** [comment]
> Do you think it´s useful to append something like `to tcp://127.0.0.1:28332` to each log entry, so that a user knows why there is an apparently duplicate entry?

Yes, makes sense in this change. Alternatively it could avoid printing the duplicate messages but not worth it.

File: test/functional/interface_zmq.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ def run_test(self):
         try:
             self.test_basic()
             self.test_reorg()
+            self.test_multiple_interfaces();
```
Comment (promag) [inline test/functional/interface_zmq.py]:
```
test/functional/interface_zmq.py:51:44: E703 statement ends with a semicolon
^---- failure generated from test/lint/lint-python.sh
```

Comment (n-thumann) [reply test/functional/interface_zmq.py]: Fixed :)

**Comment by mruddy** [comment]
Tested ACK

I tested on Ubuntu 19.10 x86_64 with zmq version 4.3.2.

I verified that the node starts when not specifying any of the zmq flags, when starting with between 1 and 3 `-zmqpubhashblock` on different addresses and/or ports, with and without `-zmqpubhashblockhwm=10000`. I verified log entries, and I also verified receiving messages with the `contrib/zmq/zmq_sub.py` subscriber.

@n-thumann side note, you might have the right skill set to review my #14687 related to zmq, if you have some spare cycles. that's a small change that the lightning guys have been requesting for a while. thanks!

**Comment by instagibbs** [comment]
concept ACK, will review

File: test/functional/interface_zmq.py [hunk 2]
```diff
@@ -204,5 +205,30 @@ def test_reorg(self):
...
+        subscribers = []
+        for x in range(2):
+            address = 'tcp://127.0.0.1:%d' % (28334 + x)
+            socket = self.ctx.socket(zmq.SUB)
+            socket.set(zmq.RCVTIMEO, 60000)
+            hashblock = ZMQSubscriber(socket, b"hashblock")
+            socket.connect(address)
+            subscribers.append({'address': address, 'socket': socket, 'hashblock': hashblock})
+
+        self.restart_node(0, ['-zmqpub%s=%s' % (subscribers[0]['hashblock'].topic.decode(), subscribers[0]['address']),
```
Comment (instagibbs) [inline test/functional/interface_zmq.py:221]: maybe compress this into:

`self.restart_node(0, ['-zmqpub%s=%s' % (subscribers[i]['hashblock'].topic.decode(), subscribers[i]['address']) for i in range(2)])`

or there-about(didn't check parenthesis lining up :) )

**Review by instagibbs — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/18309/commits/6b8d300d2560daa80d6575cb047588aec76aa5dc

Small, straight forward change. We can clean up log messages later if they're confusing.

File: test/functional/interface_zmq.py (hunk 2 above)
Comment (n-thumann) [reply test/functional/interface_zmq.py]: Thanks, applied your change slightly modified along with some other minor changes :)

**Comment by instagibbs** [comment]
re-utACK https://github.com/bitcoin/bitcoin/pull/18309/commits/ab996c2a9ecd3f75fb4621f87d9303bd18e891a2

**Comment by jonatack** [comment]
Concept ACK

**Comment by n-thumann** [comment]
Adjusted log output to new notification type of https://github.com/bitcoin/bitcoin/pull/19572.

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by laanwj** [comment]
Code review ACK e66870c5a4c2adbd30dca67d409fd5cd98697587

**Comment by instagibbs** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/18309/commits/e66870c5a4c2adbd30dca67d409fd5cd98697587

**Pull request merged**